In [651]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [652]:
data=pd.read_csv('/content/GRASIM.csv')

In [653]:
data.head()

Date  Symbol Series  Prev Close   Open   High     Low   Last   Close  \
0  2000-01-03  GRASIM     EQ      405.80  429.9  438.3  425.00  438.3  438.30   
1  2000-01-04  GRASIM     EQ      438.30  437.5  440.0  420.05  439.5  437.15   
2  2000-01-05  GRASIM     EQ      437.15  420.1  447.9  412.00  430.0  439.60   
3  2000-01-06  GRASIM     EQ      439.60  449.9  474.8  449.90  474.8  474.80   
4  2000-01-07  GRASIM     EQ      474.80  472.0  512.8  472.00  512.8  512.80   

     VWAP  Volume      Turnover  Trades  Deliverable Volume  %Deliverble  
0  432.24  101578  4.390621e+12     NaN                 NaN          NaN  
1  428.27  126117  5.401216e+12     NaN                 NaN          NaN  
2  436.10  163877  7.146660e+12     NaN                 NaN          NaN  
3  474.68   44248  2.100384e+12     NaN                 NaN          NaN  
4  512.61  188564  9.666049e+12     NaN                 NaN          NaN



*   Volume,Turnover,Trades are the metrics indicating trading activity
*   Prev Close, Open, High, Low, Last, Close: Stock price data.


*   Deliverable Volume, %Deliverble: Indicators of stock deliverability.





In [654]:
data.shape

(5306, 15)

 Analyze the indicators in the dataset that best explain the volatility and
 unpredictable nature of the stocks in the last decade.



*  Calculating volatility metrics:High-low % range,Daily Change %




In [655]:
data['Date']=pd.to_datetime(data['Date'])

In [656]:
data.head()

Date  Symbol Series  Prev Close   Open   High     Low   Last   Close  \
0 2000-01-03  GRASIM     EQ      405.80  429.9  438.3  425.00  438.3  438.30   
1 2000-01-04  GRASIM     EQ      438.30  437.5  440.0  420.05  439.5  437.15   
2 2000-01-05  GRASIM     EQ      437.15  420.1  447.9  412.00  430.0  439.60   
3 2000-01-06  GRASIM     EQ      439.60  449.9  474.8  449.90  474.8  474.80   
4 2000-01-07  GRASIM     EQ      474.80  472.0  512.8  472.00  512.8  512.80   

     VWAP  Volume      Turnover  Trades  Deliverable Volume  %Deliverble  
0  432.24  101578  4.390621e+12     NaN                 NaN          NaN  
1  428.27  126117  5.401216e+12     NaN                 NaN          NaN  
2  436.10  163877  7.146660e+12     NaN                 NaN          NaN  
3  474.68   44248  2.100384e+12     NaN                 NaN          NaN  
4  512.61  188564  9.666049e+12     NaN                 NaN          NaN

In [657]:
data['High-Low Range %'] = ((data['High'] - data['Low']) / data['Low']) * 100


In [658]:
data['Daily Change %'] = ((data['Close'] - data['Prev Close']) / data['Prev Close']) * 100


* Daily Change %: Indicates the percentage change in stock price from the previous close to the current close.
*  High-Low Range %:Represents the daily range as a percentage of the closing price.
The average daily range is approximately 3.54%, with a standard deviation of 2.55%.
Maximum observed daily range is 93.56%, suggesting highly volatile days.





In [659]:
df=data.loc[:,'Prev Close':]

In [660]:
df.head()

Prev Close   Open   High     Low   Last   Close    VWAP  Volume  \
0      405.80  429.9  438.3  425.00  438.3  438.30  432.24  101578   
1      438.30  437.5  440.0  420.05  439.5  437.15  428.27  126117   
2      437.15  420.1  447.9  412.00  430.0  439.60  436.10  163877   
3      439.60  449.9  474.8  449.90  474.8  474.80  474.68   44248   
4      474.80  472.0  512.8  472.00  512.8  512.80  512.61  188564   

       Turnover  Trades  Deliverable Volume  %Deliverble  High-Low Range %  \
0  4.390621e+12     NaN                 NaN          NaN          3.129412   
1  5.401216e+12     NaN                 NaN          NaN          4.749435   
2  7.146660e+12     NaN                 NaN          NaN          8.713592   
3  2.100384e+12     NaN                 NaN          NaN          5.534563   
4  9.666049e+12     NaN                 NaN          NaN          8.644068   

   Daily Change %  
0        8.008871  
1       -0.262377  
2        0.560448  
3        8.007279  
4        8.003370

In [661]:
new_df=data[(data['Date']>'2014-12-01')& (data['Date']<'2024-12-02')]#dataset of last decade

In [662]:
new_df.head()

Date  Symbol Series  Prev Close     Open     High      Low    Last  \
3719 2014-12-02  GRASIM     EQ     3505.80  3511.00  3529.45  3492.05  3510.0   
3720 2014-12-03  GRASIM     EQ     3511.50  3517.00  3538.35  3471.05  3477.0   
3721 2014-12-04  GRASIM     EQ     3479.05  3494.35  3496.00  3450.00  3474.0   
3722 2014-12-05  GRASIM     EQ     3473.90  3495.00  3526.00  3476.00  3489.9   
3723 2014-12-08  GRASIM     EQ     3484.60  3467.10  3512.90  3410.05  3418.0   

        Close     VWAP  Volume      Turnover  Trades  Deliverable Volume  \
3719  3511.50  3509.47   30773  1.079969e+13  2077.0             18409.0   
3720  3479.05  3500.05   37812  1.323438e+13  3790.0             21289.0   
3721  3473.90  3469.16   36533  1.267387e+13  3737.0             14436.0   
3722  3484.60  3504.93   29021  1.017165e+13  2708.0             12832.0   
3723  3437.90  3459.96   27023  9.349856e+12  2871.0             12698.0   

      %Deliverble  High-Low Range %  Daily Change %  
3719       0.5982          1.071004        0.162588  
3720       0.5630          1.938895       -0.924107  
3721       0.3951          1.333333       -0.148029  
3722       0.4422          1.438435        0.308011  
3723       0.4699          3.016085       -1.340183

In [663]:
df=new_df.loc[:,'Prev Close':]

In [664]:
df.corr()

Prev Close      Open      High       Low      Last  \
Prev Close            1.000000  0.997293  0.997186  0.997145  0.997014   
Open                  0.997293  1.000000  0.999869  0.999845  0.999694   
High                  0.997186  0.999869  1.000000  0.999813  0.999867   
Low                   0.997145  0.999845  0.999813  1.000000  0.999855   
Last                  0.997014  0.999694  0.999867  0.999855  1.000000   
Close                 0.997039  0.999717  0.999882  0.999870  0.999989   
VWAP                  0.997156  0.999841  0.999933  0.999938  0.999941   
Volume               -0.507464 -0.508628 -0.505920 -0.509718 -0.507758   
Turnover             -0.372655 -0.373840 -0.370585 -0.375838 -0.372947   
Trades               -0.568697 -0.569451 -0.566324 -0.571548 -0.568622   
Deliverable Volume   -0.380149 -0.381188 -0.379925 -0.382416 -0.381543   
%Deliverble           0.250052  0.250103  0.248084  0.250325  0.248416   
High-Low Range %     -0.268091 -0.269666 -0.263539 -0.276161 -0.269240   
Daily Change %       -0.040255  0.012904  0.018757  0.018943  0.024228   

                       Close      VWAP    Volume  Turnover    Trades  \
Prev Close          0.997039  0.997156 -0.507464 -0.372655 -0.568697   
Open                0.999717  0.999841 -0.508628 -0.373840 -0.569451   
High                0.999882  0.999933 -0.505920 -0.370585 -0.566324   
Low                 0.999870  0.999938 -0.509718 -0.375838 -0.571548   
Last                0.999989  0.999941 -0.507758 -0.372947 -0.568622   
Close               1.000000  0.999957 -0.507613 -0.372683 -0.568472   
VWAP                0.999957  1.000000 -0.507710 -0.372975 -0.568702   
Volume             -0.507613 -0.507710  1.000000  0.928349  0.884384   
Turnover           -0.372683 -0.372975  0.928349  1.000000  0.869140   
Trades             -0.568472 -0.568702  0.884384  0.869140  1.000000   
Deliverable Volume -0.381431 -0.381425  0.816976  0.841834  0.715388   
%Deliverble         0.248503  0.248836 -0.257817 -0.130319 -0.184986   
High-Low Range %   -0.269352 -0.269932  0.493416  0.400573  0.513714   
Daily Change %      0.024017  0.020653  0.046499  0.048108  0.046768   

                    Deliverable Volume  %Deliverble  High-Low Range %  \
Prev Close                   -0.380149     0.250052         -0.268091   
Open                         -0.381188     0.250103         -0.269666   
High                         -0.379925     0.248084         -0.263539   
Low                          -0.382416     0.250325         -0.276161   
Last                         -0.381543     0.248416         -0.269240   
Close                        -0.381431     0.248503         -0.269352   
VWAP                         -0.381425     0.248836         -0.269932   
Volume                        0.816976    -0.257817          0.493416   
Turnover                      0.841834    -0.130319          0.400573   
Trades                        0.715388    -0.184986          0.513714   
Deliverable Volume            1.000000     0.138440          0.325865   
%Deliverble                   0.138440     1.000000         -0.218801   
High-Low Range %              0.325865    -0.218801          1.000000   
Daily Change %               -0.020595    -0.063517         -0.010391   

                    Daily Change %  
Prev Close               -0.040255  
Open                      0.012904  
High                      0.018757  
Low                       0.018943  
Last                      0.024228  
Close                     0.024017  
VWAP                      0.020653  
Volume                    0.046499  
Turnover                  0.048108  
Trades                    0.046768  
Deliverable Volume       -0.020595  
%Deliverble              -0.063517  
High-Low Range %         -0.010391  
Daily Change %            1.000000

Correlation Analysis
High-Low Range %--



* Trades:Positive Correlation(0.513) between High-low range(%) and Trades
*  Volume:Positive Correlation(0.4934).Increase in Volume lead to increase in Trades.  

*  Turnover:Positive Correlation(0.400).Increase in Turnover lead to increase in High-Low Range
*  Deliverable%:Negative Correlation(-0.21) between Deliverable Volume and High ow Range%








Daily Change % (Price Change Volatility):
* Turnover:Positive correlation(-0.048) suggest a slight relationship between Turnover and Daily Change%.
* Trades: Positive correlation(-0.046) suggest a slight relationship between Turnover and Daily Change%.

* Volume:Positive Correlation(0.046) suggest a positive relationship
*  Deliverable Volume:Positive Correlation(-0.020)suggest a positive relationship between Deliverable Volume and Daily Change%.
   

Insights

1.  Trades and Volume are critical indicators of intraday volatility.
2.  Turnover slightly affects price volatility but has a stronger impact on intraday volatility.
3.%Deliverable inversely correlates with both volatility measures, potentially indicating that stocks with more speculative trading (lower deliverability) are more volatile

In [665]:
data.isnull().sum()

Date                     0
Symbol                   0
Series                   0
Prev Close               0
Open                     0
High                     0
Low                      0
Last                     0
Close                    0
VWAP                     0
Volume                   0
Turnover                 0
Trades                2850
Deliverable Volume     509
%Deliverble            509
High-Low Range %         0
Daily Change %           0
dtype: int64

In [666]:
for col in data.columns:
  if data[col].dtype=='int64' or data[col].dtype=='float64':
    data[col]=data[col].fillna(data[col].mean())

In [667]:
data.isna().sum()

Date                  0
Symbol                0
Series                0
Prev Close            0
Open                  0
High                  0
Low                   0
Last                  0
Close                 0
VWAP                  0
Volume                0
Turnover              0
Trades                0
Deliverable Volume    0
%Deliverble           0
High-Low Range %      0
Daily Change %        0
dtype: int64

#Choose promising stock based on your analysis for your portfolio

In [668]:
data

Date  Symbol Series  Prev Close     Open    High      Low     Last  \
0    2000-01-03  GRASIM     EQ      405.80   429.90   438.3   425.00   438.30   
1    2000-01-04  GRASIM     EQ      438.30   437.50   440.0   420.05   439.50   
2    2000-01-05  GRASIM     EQ      437.15   420.10   447.9   412.00   430.00   
3    2000-01-06  GRASIM     EQ      439.60   449.90   474.8   449.90   474.80   
4    2000-01-07  GRASIM     EQ      474.80   472.00   512.8   472.00   512.80   
...         ...     ...    ...         ...      ...     ...      ...      ...   
5301 2021-04-26  GRASIM     EQ     1278.10  1285.00  1321.0  1285.00  1320.00   
5302 2021-04-27  GRASIM     EQ     1312.35  1320.00  1339.2  1308.00  1338.00   
5303 2021-04-28  GRASIM     EQ     1331.20  1337.45  1349.0  1321.15  1347.25   
5304 2021-04-29  GRASIM     EQ     1341.15  1354.40  1369.5  1332.30  1352.00   
5305 2021-04-30  GRASIM     EQ     1350.70  1342.00  1407.4  1333.00  1401.10   

        Close     VWAP   Volume      Turnover        Trades  \
0      438.30   432.24   101578  4.390621e+12  21712.739821   
1      437.15   428.27   126117  5.401216e+12  21712.739821   
2      439.60   436.10   163877  7.146660e+12  21712.739821   
3      474.80   474.68    44248  2.100384e+12  21712.739821   
4      512.80   512.61   188564  9.666049e+12  21712.739821   
...       ...      ...      ...           ...           ...   
5301  1312.35  1303.61  1233546  1.608066e+14  36437.000000   
5302  1331.20  1323.07  1236334  1.635755e+14  48465.000000   
5303  1341.15  1337.11  1283746  1.716511e+14  33235.000000   
5304  1350.70  1348.65  1793007  2.418143e+14  52053.000000   
5305  1401.75  1386.13  3261678  4.521116e+14  87943.000000   

      Deliverable Volume  %Deliverble  High-Low Range %  Daily Change %  
0           2.109953e+05     0.495072          3.129412        8.008871  
1           2.109953e+05     0.495072          4.749435       -0.262377  
2           2.109953e+05     0.495072          8.713592        0.560448  
3           2.109953e+05     0.495072          5.534563        8.007279  
4           2.109953e+05     0.495072          8.644068        8.003370  
...                  ...          ...               ...             ...  
5301        3.832380e+05     0.310700          2.801556        2.679759  
5302        4.475530e+05     0.362000          2.385321        1.436355  
5303        6.183860e+05     0.481700          2.108012        0.747446  
5304        8.721790e+05     0.486400          2.792164        0.712075  
5305        1.445017e+06     0.443000          5.581395        3.779522  

[5306 rows x 17 columns]

In [669]:
data.columns

Index(['Date', 'Symbol', 'Series', 'Prev Close', 'Open', 'High', 'Low', 'Last',
       'Close', 'VWAP', 'Volume', 'Turnover', 'Trades', 'Deliverable Volume',
       '%Deliverble', 'High-Low Range %', 'Daily Change %'],
      dtype='object')

In [670]:
# Calculate summary statistics by stock
stock_analysis = data.groupby('Symbol').agg({
    'Daily Change %': ['mean', 'std', lambda x: (x > 0).sum() / len(x)],  # Mean, Std Dev, and Positive Return Frequency
    'High-Low Range %': ['mean'],                                           # Mean volatility
    'Volume': ['mean'],                                                   # Mean trading volume
    '%Deliverble': ['mean'],                                              # Mean deliverable percentage
    'VWAP': ['mean']                                                      # Average VWAP
})

# Flatten column names
stock_analysis.columns = [
    'Mean Return (%)', 'Return Std Dev (%)', 'Positive Return Frequency',
    'Mean Volatility (%)', 'Mean Volume', 'Mean Deliverable (%)', 'Mean VWAP'
]

# Define thresholds or scores for the criteria
stock_analysis['Consistency Score'] = stock_analysis['Positive Return Frequency'] * 100  # Positive return frequency as a percentage
stock_analysis['Moderate Volatility Score'] = -abs(stock_analysis['Mean Volatility (%)'] - stock_analysis['Mean Volatility (%)'].median())
stock_analysis['High Volume Score'] = stock_analysis['Mean Volume'] / stock_analysis['Mean Volume'].max() * 100  # Volume as a percentage of the max volume
stock_analysis['Trend Score'] = stock_analysis['Mean Deliverable (%)'] + stock_analysis['Mean VWAP']  # Simplified trend measure

# Combine scores for overall ranking
stock_analysis['Overall Score'] = (
    stock_analysis['Consistency Score'] +
    stock_analysis['Moderate Volatility Score'] +
    stock_analysis['High Volume Score'] +
    stock_analysis['Trend Score']
)

# Rank stocks based on the overall score
stock_analysis['Rank'] = stock_analysis['Overall Score'].rank(ascending=False)

# Select the top stocks based on overall rank
top_consistent_stocks = stock_analysis.sort_values('Rank').head(5).reset_index()

# Display the top stocks
print(top_consistent_stocks[['Symbol', 'Mean Return (%)', 'Positive Return Frequency',
                             'Mean Volatility (%)', 'Mean Volume',
                             'Mean Deliverable (%)', 'Mean VWAP', 'Overall Score']])


   Symbol  Mean Return (%)  Positive Return Frequency  Mean Volatility (%)  \
0  GRASIM         0.063499                   0.503015             3.456374   

     Mean Volume  Mean Deliverable (%)    Mean VWAP  Overall Score  
0  481077.368262              0.495072  1753.952433     1904.74905  


In [671]:
data.head()

Date  Symbol Series  Prev Close   Open   High     Low   Last   Close  \
0 2000-01-03  GRASIM     EQ      405.80  429.9  438.3  425.00  438.3  438.30   
1 2000-01-04  GRASIM     EQ      438.30  437.5  440.0  420.05  439.5  437.15   
2 2000-01-05  GRASIM     EQ      437.15  420.1  447.9  412.00  430.0  439.60   
3 2000-01-06  GRASIM     EQ      439.60  449.9  474.8  449.90  474.8  474.80   
4 2000-01-07  GRASIM     EQ      474.80  472.0  512.8  472.00  512.8  512.80   

     VWAP  Volume      Turnover        Trades  Deliverable Volume  \
0  432.24  101578  4.390621e+12  21712.739821       210995.283927   
1  428.27  126117  5.401216e+12  21712.739821       210995.283927   
2  436.10  163877  7.146660e+12  21712.739821       210995.283927   
3  474.68   44248  2.100384e+12  21712.739821       210995.283927   
4  512.61  188564  9.666049e+12  21712.739821       210995.283927   

   %Deliverble  High-Low Range %  Daily Change %  
0     0.495072          3.129412        8.008871  
1     0.495072          4.749435       -0.262377  
2     0.495072          8.713592        0.560448  
3     0.495072          5.534563        8.007279  
4     0.495072          8.644068        8.003370

In [672]:
data

Date  Symbol Series  Prev Close     Open    High      Low     Last  \
0    2000-01-03  GRASIM     EQ      405.80   429.90   438.3   425.00   438.30   
1    2000-01-04  GRASIM     EQ      438.30   437.50   440.0   420.05   439.50   
2    2000-01-05  GRASIM     EQ      437.15   420.10   447.9   412.00   430.00   
3    2000-01-06  GRASIM     EQ      439.60   449.90   474.8   449.90   474.80   
4    2000-01-07  GRASIM     EQ      474.80   472.00   512.8   472.00   512.80   
...         ...     ...    ...         ...      ...     ...      ...      ...   
5301 2021-04-26  GRASIM     EQ     1278.10  1285.00  1321.0  1285.00  1320.00   
5302 2021-04-27  GRASIM     EQ     1312.35  1320.00  1339.2  1308.00  1338.00   
5303 2021-04-28  GRASIM     EQ     1331.20  1337.45  1349.0  1321.15  1347.25   
5304 2021-04-29  GRASIM     EQ     1341.15  1354.40  1369.5  1332.30  1352.00   
5305 2021-04-30  GRASIM     EQ     1350.70  1342.00  1407.4  1333.00  1401.10   

        Close     VWAP   Volume      Turnover        Trades  \
0      438.30   432.24   101578  4.390621e+12  21712.739821   
1      437.15   428.27   126117  5.401216e+12  21712.739821   
2      439.60   436.10   163877  7.146660e+12  21712.739821   
3      474.80   474.68    44248  2.100384e+12  21712.739821   
4      512.80   512.61   188564  9.666049e+12  21712.739821   
...       ...      ...      ...           ...           ...   
5301  1312.35  1303.61  1233546  1.608066e+14  36437.000000   
5302  1331.20  1323.07  1236334  1.635755e+14  48465.000000   
5303  1341.15  1337.11  1283746  1.716511e+14  33235.000000   
5304  1350.70  1348.65  1793007  2.418143e+14  52053.000000   
5305  1401.75  1386.13  3261678  4.521116e+14  87943.000000   

      Deliverable Volume  %Deliverble  High-Low Range %  Daily Change %  
0           2.109953e+05     0.495072          3.129412        8.008871  
1           2.109953e+05     0.495072          4.749435       -0.262377  
2           2.109953e+05     0.495072          8.713592        0.560448  
3           2.109953e+05     0.495072          5.534563        8.007279  
4           2.109953e+05     0.495072          8.644068        8.003370  
...                  ...          ...               ...             ...  
5301        3.832380e+05     0.310700          2.801556        2.679759  
5302        4.475530e+05     0.362000          2.385321        1.436355  
5303        6.183860e+05     0.481700          2.108012        0.747446  
5304        8.721790e+05     0.486400          2.792164        0.712075  
5305        1.445017e+06     0.443000          5.581395        3.779522  

[5306 rows x 17 columns]

 Feature engineer and build machine learning solutions for the business
 Requirements

In [673]:
data['Cumulative_Return'] = data['Close'] / data['Close'].iloc[0] - 1  #cumilative returns over a period


In [674]:
data['Volume_Change'] = data['Volume'].pct_change()


In [675]:
#Moving averages
data['SMA_50'] = data['Close'].rolling(window=50).mean()
data['SMA_200'] = data['Close'].rolling(window=200).mean()


In [676]:
data['Volatility'] = data['Daily Change %'].rolling(window=20).std()


In [677]:
data.head()

Date  Symbol Series  Prev Close   Open   High     Low   Last   Close  \
0 2000-01-03  GRASIM     EQ      405.80  429.9  438.3  425.00  438.3  438.30   
1 2000-01-04  GRASIM     EQ      438.30  437.5  440.0  420.05  439.5  437.15   
2 2000-01-05  GRASIM     EQ      437.15  420.1  447.9  412.00  430.0  439.60   
3 2000-01-06  GRASIM     EQ      439.60  449.9  474.8  449.90  474.8  474.80   
4 2000-01-07  GRASIM     EQ      474.80  472.0  512.8  472.00  512.8  512.80   

     VWAP  ...        Trades  Deliverable Volume  %Deliverble  \
0  432.24  ...  21712.739821       210995.283927     0.495072   
1  428.27  ...  21712.739821       210995.283927     0.495072   
2  436.10  ...  21712.739821       210995.283927     0.495072   
3  474.68  ...  21712.739821       210995.283927     0.495072   
4  512.61  ...  21712.739821       210995.283927     0.495072   

   High-Low Range %  Daily Change %  Cumulative_Return  Volume_Change  SMA_50  \
0          3.129412        8.008871           0.000000            NaN     NaN   
1          4.749435       -0.262377          -0.002624       0.241578     NaN   
2          8.713592        0.560448           0.002966       0.299405     NaN   
3          5.534563        8.007279           0.083276      -0.729993     NaN   
4          8.644068        8.003370           0.169975       3.261526     NaN   

   SMA_200  Volatility  
0      NaN         NaN  
1      NaN         NaN  
2      NaN         NaN  
3      NaN         NaN  
4      NaN         NaN  

[5 rows x 22 columns]

In [678]:
data['Target_Close'] = data['Close'].shift(-1)


In [679]:
data.head()

Date  Symbol Series  Prev Close   Open   High     Low   Last   Close  \
0 2000-01-03  GRASIM     EQ      405.80  429.9  438.3  425.00  438.3  438.30   
1 2000-01-04  GRASIM     EQ      438.30  437.5  440.0  420.05  439.5  437.15   
2 2000-01-05  GRASIM     EQ      437.15  420.1  447.9  412.00  430.0  439.60   
3 2000-01-06  GRASIM     EQ      439.60  449.9  474.8  449.90  474.8  474.80   
4 2000-01-07  GRASIM     EQ      474.80  472.0  512.8  472.00  512.8  512.80   

     VWAP  ...  Deliverable Volume  %Deliverble  High-Low Range %  \
0  432.24  ...       210995.283927     0.495072          3.129412   
1  428.27  ...       210995.283927     0.495072          4.749435   
2  436.10  ...       210995.283927     0.495072          8.713592   
3  474.68  ...       210995.283927     0.495072          5.534563   
4  512.61  ...       210995.283927     0.495072          8.644068   

   Daily Change %  Cumulative_Return  Volume_Change  SMA_50  SMA_200  \
0        8.008871           0.000000            NaN     NaN      NaN   
1       -0.262377          -0.002624       0.241578     NaN      NaN   
2        0.560448           0.002966       0.299405     NaN      NaN   
3        8.007279           0.083276      -0.729993     NaN      NaN   
4        8.003370           0.169975       3.261526     NaN      NaN   

   Volatility  Target_Close  
0         NaN        437.15  
1         NaN        439.60  
2         NaN        474.80  
3         NaN        512.80  
4         NaN        501.95  

[5 rows x 23 columns]

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
import matplotlib.pyplot as plt


# Ensure the Date column is in datetime format
data['Date'] = pd.to_datetime(data['Date'])


# Feature Engineering
data['Daily_Return'] = data['Close'].pct_change()
data['SMA_20'] = data['Close'].rolling(window=20).mean()
data['SMA_50'] = data['Close'].rolling(window=50).mean()
data['Volatility'] = data['Close'].rolling(window=20).std()

# Drop rows with NaN values generated from rolling calculations
data.dropna(inplace=True)
# Drop the last row with NaN target
data = data.dropna()

# Define Features (X) and Target (y)
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'Daily_Return', 'SMA_20', 'SMA_50', 'Volatility']
X = data[features]
y = data['Target_Close']

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train Random Forest Regressor
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Make Predictions
y_pred = rf_model.predict(X_test)

# Evaluate the Model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R2 Score: {r2}")

# Plot Actual vs Predicted Prices
plt.figure(figsize=(10, 6))
plt.plot(y_test.values, label="Actual Prices", color="blue")
plt.plot(y_pred, label="Predicted Prices", color="orange")
plt.title("Actual vs Predicted Prices")
plt.xlabel("Time")
plt.ylabel("Stock Price")
plt.legend()
plt.show()


In [ ]:
data.head()

In [ ]:
features=['Date','Close']

In [ ]:
my_data=data[features]

In [ ]:
my_data['Date']=pd.to_datetime(my_data['Date'])

In [ ]:
my_data

In [ ]:
my_data.index=my_data['Date']

In [ ]:
del my_data['Date']

In [ ]:
my_data.plot()

In [ ]:
std_log=my_data.rolling(window=4).std()

In [ ]:
mean_log=my_data.rolling(window=4).mean()

In [ ]:
plt.plot(my_data,color='blue',label='Original Data')
plt.plot(mean_log,color='black',label='Rolling Mean')
plt.plot(std_log,color='red',label='Rolling Standard Deviation')
plt.legend(loc='best')

In [ ]:
from statsmodels.tsa.stattools  import adfuller

In [ ]:
#Augmented DickyFuller Test
result=adfuller(my_data['Close'])
keys=['test_statistics','p_value',"num_lags","num_observations","critical_values","AIC"]

In [ ]:
for key,value in zip(keys,result):
  print(f"{key}:{value}")


In [ ]:
import numpy as np

In [ ]:
first_log=np.log(my_data)

In [ ]:
first_log.dropna()

In [ ]:
first_log.plot()

In [ ]:
new_data=first_log-mean_log

In [ ]:
mean_log=new_data.rolling(window=12).mean()
std_log=new_data.rolling(window=12).std()

plt.plot(new_data,color='blue',label='Original data')
plt.plot(mean_log,color='red',label='Mean data')
plt.plot(std_log,color='black',label='Standard Deviation')

plt.legend(loc='best')
plt.title("Rolling mean and Standard deviation for log_scale data")

In [ ]:
new_data

In [ ]:
from statsmodels.tsa.stattools import acf,pacf
from statsmodels.graphics.tsaplots import plot_acf,plot_pacf

In [ ]:
acf_values=acf(new_data)
plot_acf(new_data)
pacf_values=pacf(new_data,nlags=10)
plot_pacf(new_data)

In [ ]:
from statsmodels.tsa.arima.model import ARIMA

In [ ]:
train=new_df.iloc[:120]['Close']

In [ ]:
test=new_df.iloc[120:]

In [ ]:
model=ARIMA(train,order=(1,0,2))

In [ ]:
model_fit=model.fit()

In [ ]:
print(model_fit.summary())

In [ ]:
new_data['predicted']=model_fit.predict(start=len(train),end=len(train) + len(test)-1,dynamic=True)
new_data[['Close','predicted']].plot()

In [ ]:
from statsmodels.tsa.statespace.sarimax import SARIMAX

In [ ]:
model=SARIMAX(train,order=(1,0,2),seasonal_order=(1,0,2,12))
model_fit=model.fit()

In [ ]:
new_data['predicted']=model_fit.predict(start=len(train),end=len(train) + len(test)-1,dynamic=True)
new_data[['Close','predicted']].plot()

In [ ]:
forecast=model_fit.forecast(steps=60)
plt.figure(figsize=(10,6))
plt.plot(new_data.index,new_data['Close'],label='Original Data')

In [ ]:
plt.plot(forecast.index,forecast,label='Forecast',linestyle='--')
plt.xlabel("Date")
plt.ylabel("Close")
plt.title("Forecasting Stock Prices for the next five years")
plt.legend()